helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -874,7 +926,31 @@ public int compare(ResourcePriority priority1, ResourcePriority priority2) {
     }
   }
 
-  // Compare partitions according following standard:
+  private class MessagePriorityComparator implements Comparator<Message> {
+    private Map<String, Integer> _preferenceInstanceMap;
+    private Map<String, Integer> _statePriorityMap;
+
+    MessagePriorityComparator(List<String> preferenceList, Map<String, Integer> statePriorityMap) {
+      // Get instance -> priority map.
+      _preferenceInstanceMap = IntStream.range(0, preferenceList.size())
+          .boxed()
+          .collect(Collectors.toMap(preferenceList::get, index -> index));
+      _statePriorityMap = statePriorityMap;
+    }
+
+    @Override
+    public int compare(Message m1, Message m2) {
+      //Compare rules:
+      //     1. Higher target state has higher priority.
+      //     2. If target state is same, range it as preference list order.
+      if (m1.getToState().equals(m2.getToState())) {
+        return _preferenceInstanceMap.get(m1.getTgtName()).compareTo(_preferenceInstanceMap.get(m2.getTgtName()));
```

Comment:
A message might target an instance that does not exist in the preference list.

Reply:
For safety and future extension, maybe it is a good idea to sort the messages by name as the last resort.

Reply:
I can make the sort check for that. But we dont need sort it by name. Because if the instance does not exist in preference list, it means the message for this instance should be downward or something. Then sort by state of target state will make it last priority.

Reply:
The reason for sorting by naming is more deterministic. I agree that there is no solid reason that we have to sort by the names. But it doesn't hurt and can help to avoid any randomness which might be annoying.

BTW, the first problem that I mentioned is not related to sorting. I believe an NPE will be thrown if you don't do a null check.